iree/compiler/Dialect/HAL/IR/HALTypes.cpp

```diff
@@ -58,6 +60,243 @@ static LogicalResult parseEnumAttr(DialectAsmParser &parser, StringRef attrName,
   return success();
 }
 
+template <typename AttrType>
+static LogicalResult parseOptionalEnumAttr(DialectAsmParser &parser,
+                                           StringRef attrName, AttrType &attr) {
+  if (succeeded(parser.parseOptionalQuestion())) {
+    // Special case `?` to indicate any/none/undefined/etc.
+    attr = AttrType::get(parser.getBuilder().getContext(), 0);
+    return success();
+  }
+  return parseEnumAttr<AttrType>(parser, attrName, attr);
+}
+
+static LogicalResult parseMemoryType(DialectAsmParser &parser,
+                                     Attribute &attr) {
+  if (succeeded(parser.parseOptionalQuestion())) {
+    attr = parser.getBuilder().getI32IntegerAttr(0);
+    return success();
+  }
+
+  StringRef fullString;
+  if (succeeded(parser.parseOptionalString(&fullString))) {
+    auto symbolized = symbolizeEnum<MemoryTypeBitfield>(fullString);
+    if (!symbolized.hasValue()) {
+      return parser.emitError(parser.getCurrentLocation())
+             << "failed to parse memory type enum value";
+    }
+    attr = parser.getBuilder().getI32IntegerAttr(
+        static_cast<int32_t>(symbolized.getValue()));
+    return success();
+  }
+
+  StringRef shortString;
+  if (failed(parser.parseKeyword(&shortString))) {
+    return parser.emitError(parser.getCurrentLocation())
+           << "failed to find memory type short string";
+  }
+  MemoryTypeBitfield memoryType = MemoryTypeBitfield::None;
+  for (char c : shortString) {
+    switch (c) {
+      case 'T':
+        memoryType = memoryType | MemoryTypeBitfield::Transient;
+        break;
+      case 'h':
+        memoryType = memoryType | MemoryTypeBitfield::HostVisible;
+        break;
```

Comment:
Ah, this answers my question about the single character comments, I think. Could still add more context to the .td though.

Reply:
Yeah this is pulling some stuff from the future where buffer types use these - I'll continue refining them when we can produce IR that uses them and make some judgements about what makes sense. This first pass was done during my previous buffer allocation work when having 100k `!hal.buffer<"HostVisible|DeviceLocal|DeviceVisible", "Constant|Transfer|Mapping", "Read|Write">` made my eyes bleed vs `!hal.buffer<hD:CTM:RW>`.

Reply:
Ok cool. Still seems worth a comment in the `.td`, perhaps with just that example you included

Reply:
(I still think a comment or some more documentation on the short string versions would be useful, but that can be added as this gets refined)